File: DM&ADS/sorts/G.java
import java.io.*;
import java.util.*;

public class Style2 {
    static int min(int a1,int a2,int a3,int a4){
        return Math.min(Math.min(a1,a2),Math.min(a3,a4));
    }
    static int max(int a1,int a2,int a3,int a4){
        return Math.max(Math.max(a1,a2),Math.max(a3,a4));
    }


    void solve(PrintWriter out) throws IOException {
        FastScanner in = new FastScanner("style.in");
        int n1 = in.nextInt();
        int[] ar1 = new int[n1];
        for (int i = 0; i < n1; i++) {
            ar1[i] = in.nextInt();
        }
        int n2 = in.nextInt();
        int[] ar2 = new int[n2];
        for (int i = 0; i < n2; i++) {
            ar2[i] = in.nextInt();
        }
        int n3 = in.nextInt();
        int[] ar3 = new int[n3];
        for (int i = 0; i < n3; i++) {
            ar3[i] = in.nextInt();
        }
        int n4 = in.nextInt();
        int[] ar4 = new int[n4];
        for (int i = 0; i < n4; i++) {
            ar4[i] = in.nextInt();
        }
        Arrays.sort(ar1);
        Arrays.sort(ar2);
        Arrays.sort(ar3);
        Arrays.sort(ar4);
        int i1=0;
        int i2=0;
        int i3=0;
        int i4=0;
        int res1=0,res2=0,res3=0,res4=0;
        int minRes=Integer.MAX_VALUE;
        int abs=max(ar1[i1],ar2[i2],ar3[i3],ar4[i4])-min(ar1[i1],ar2[i2],ar3[i3],ar4[i4]);
        if (abs<minRes){
            minRes=abs;
            res1=ar1[i1];
            res2=ar2[i2];
            res3=ar3[i3];
            res4=ar4[i4];
        }
        while (true){
            int min=min(ar1[i1],ar2[i2],ar3[i3],ar4[i4]);
            if ( ar1[i1]==min) i1++;
            else if (ar2[i2]==min) i2++;
            else if (ar3[i3]==min) i3++;
            else if (ar4[i4]==min) i4++;
            if (!(i1<n1 && i2<n2 && i3<n3 && i4<n4) ){
                break;
            }
            abs=max(ar1[i1],ar2[i2],ar3[i3],ar4[i4])-min(ar1[i1],ar2[i2],ar3[i3],ar4[i4]);
            if (abs<minRes){
                minRes=abs;
                res1=ar1[i1];
                res2=ar2[i2];
                res3=ar3[i3];
                res4=ar4[i4];
            }
        }
        out.print(res1+" "+res2+" "+res3+" "+res4);

    }


    StreamTokenizer in;
    PrintWriter out;

    void run() throws IOException {
        out = new PrintWriter("style.out");
        solve(out);
        out.flush();
    }

    public static void main(String[] args) throws IOException {
        new Style2().run();

    }

    class FastScanner {
        BufferedReader reader;
        StringTokenizer tokenizer;

        public FastScanner(String fileName) throws IOException {
            reader = new BufferedReader(new FileReader(fileName));
        }

        public String next() throws IOException {
            while (tokenizer == null || !tokenizer.hasMoreTokens()) {
                String line = reader.readLine();
                if (line == null || line.equals("")) {
                    return "";
                }
                tokenizer = new StringTokenizer(line);
            }
            return tokenizer.nextToken();
        }

        public int nextInt() throws IOException {
            return Integer.parseInt(next());
        }

        public long nextLong() throws IOException {
            return Long.parseLong(next());
        }
    }

}

File: DM&ADS/flows/maxflow/G.kt
package algoLab10

import java.io.*
import java.lang.Integer.min
import java.lang.NullPointerException
import java.util.*
import java.util.stream.Collectors
import kotlin.collections.ArrayList


object G {
    const val fileName = ""

    //================================================================================
    class Edge(
            val from: Int,
            val to: Int,
            val cap: Int,
            val i: Int = 0,
            val j: Int = 0
    ) {
        var backEdge: Edge? = null
        var flow: Int = 0


    }


    var N: Int = 0
    var M: Int = 0
    var S: Int = 0
    var T: Int = 0
    lateinit var graph: Array<ArrayList<Edge>>
    lateinit var table: Array<Array<Char>>
    lateinit var visited: Array<Boolean>
    val edges = ArrayList<Edge>()
    // i - in , N * M + i - out
    private fun addVertexEdge(i: Int, j: Int, cap: Int) {
        val from = i * M + j
        val to = N * M + i * M + j
        val edge = Edge(from, to, cap, i, j)
        edges.add(edge)
        val backEdge = Edge(to, from, 0, i, j)
        edge.backEdge = backEdge
        backEdge.backEdge = edge
        graph[from].add(edge)
        graph[to].add(backEdge)

    }

    const val BIG_CAP = Short.MAX_VALUE.toInt()

    private fun add(from: Int, to: Int, cap: Int, backCap: Int = 0) {
        val edge = Edge(from, to, cap)
        val backEdge = Edge(to, from, backCap)
        edge.backEdge = backEdge
        backEdge.backEdge = edge
        graph[from].add(edge)
        graph[to].add(backEdge)
    }

    private fun run() {
        N = IO.nextInt()
        M = IO.nextInt()
        graph = Array(2 * N * M) { ArrayList<Edge>() }
        table = Array(N) { Array(M) { ' ' } }
        for (i in 0 until N) {
            val line = IO.nextLine()
            for (j in 0 until line.length) {
                val ch = line[j]
                table[i][j] = ch
                when (ch) {
                    // '#' -> addVertexEdge(i, j, 0)
                    '.' -> addVertexEdge(i, j, 1)
                    '-' -> addVertexEdge(i, j, BIG_CAP)
                    'A' -> {
                        S = N * M + i * M + j
                    }
                    'B' -> {
                        T = i * M + j
                    }
                }
            }

        }
        for (i in 0 until N) {
            for (j in 0 until M) {
                val inp = i * M + j
                val out = N * M + i * M + j
                if (i != N - 1) {
                    add(out, M + inp, BIG_CAP)
                    add(out + M, inp, BIG_CAP)
                }
                if (j != M - 1) {
                    add(out, 1 + inp, BIG_CAP)
                    add(out + 1, inp, BIG_CAP)
                }
            }
        }
        var maxFlow = 0
        while (true) {
            visited = Array(N * M * 2) { false }
            val res = dfsFlow(S, Short.MAX_VALUE.toInt())
            maxFlow += res
            if (res == 0) break
        }
        if (maxFlow >= BIG_CAP) {
            IO.println(-1)
            return
        }
        if (maxFlow == 0) {
            IO.println(0)
            table.forEach { i ->
                i.forEach { j ->
                    IO.print(j)
                }
                IO.println("")
            }
            return
        }
        visited = Array(N * M * 2) { false }
        dfsCut(S)
        val cut = edges.stream().filter { visited[it.to] != visited[it.from] }.collect(Collectors.toList())
        for (edge in cut) {
            table[edge.i][edge.j] = '+'
        }
//        for (v in 0 until N * M) {
//            if (!visited[v]) continue
//            for (edge in graph[v]) {
//                if (edge.cap == 1 && edge.flow == 1 && !visited[edge.to]) {
//                    table[edge.i][edge.j] = '+'
//                    break
//                }
//            }
//
//        }
        IO.println(maxFlow)
        table.forEach { i ->
            i.forEach { j ->
                IO.print(j)
            }
            IO.println("")
        }
    }


    private fun dfsCut(from: Int) {
        if (from == T) {
            return
        }
        visited[from] = true
        for (edge in graph[from]) {
            val to = edge.to
            if (visited[to]) {
                continue
            }
            if (edge.flow < edge.cap) {
                dfsCut(to)
            }
        }
    }


    private fun dfsFlow(from: Int, c: Int): Int {
        if (from == T) {
            return c
        }
        if (visited[from]) {
            return 0
        }
        visited[from] = true
        for (edge in graph[from]) {
            val to = edge.to
            if (!visited[to] && edge.flow < edge.cap) {
                val left = dfsFlow(to, min(c, edge.cap - edge.flow))
                if (left > 0) {
                    edge.flow += left
                    edge.backEdge!!.flow -= left
                    return left
                }

            }
        }
        return 0

    }
    //==================================================================================

    @JvmStatic
    fun main(args: Array<String>) {
        this.run()
        IO.writer.flush()
        IO.reader.close()

    }

    object IO {
        val reader: BufferedReader
        val writer: PrintWriter
        private var tokenizer: StringTokenizer? = null

        init {
            if (fileName.isEmpty()) {
                reader = BufferedReader(InputStreamReader(System.`in`))
                writer = PrintWriter(System.out)
            } else {
                reader = BufferedReader(FileReader("$fileName.in"))
                writer = PrintWriter(FileWriter("$fileName.out"))
            }
        }

        fun next(): String {
            while (tokenizer == null || !tokenizer!!.hasMoreTokens()) {
                val line = reader.readLine() ?: throw NullPointerException()
                if (line == "") {
                    return ""
                }
                tokenizer = StringTokenizer(line)
            }
            return tokenizer!!.nextToken()
        }

        fun nextLine(): String {
            return reader.readLine()
        }

        fun nextInt(): Int {
            return next().toInt()
        }

        fun nextDouble(): Double {
            return next().toDouble()
        }

        fun nextLong(): Long {
            return next().toLong()
        }

        fun <T> print(obj: T) {
            writer.print(obj)
        }

        fun <T> println(obj: T) {
            writer.println(obj)
        }
    }
}